Advanced/Play Asset Delivery/Assets/PlayAssetDelivery/Editor/PlayAssetDeliveryBuildProcessor.cs

```diff
@@ -0,0 +1,97 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using UnityEditor;
+using UnityEditor.Build;
+using UnityEditor.Build.Reporting;
+using UnityEditor.Callbacks;
+using UnityEngine;
+
+namespace AddressablesPlayAssetDelivery.Editor
+{
+    /// <summary>
+    /// Moves custom asset pack data from their default build location <see cref="BuildScriptPlayAssetDelivery"/> to their correct player build data location.
+    /// For an Android App Bundle, bundles assigned to a custom asset pack must be located in their {asset pack name}.androidpack directory in the Assets folder.
+    /// The 'CustomAssetPacksData.json' file is also moved to StreamingAssets.
+    ///
+    /// This script executes before the <see cref="AddressablesPlayerBuildProcessor"/> which moves all Addressables data to StreamingAssets.
+    /// </summary>
+    public class PlayAssetDeliveryBuildProcessor : IPreprocessBuildWithReport
+    {
+        /// <summary>
+        /// Returns the player build processor callback order.
+        /// </summary>
+        public int callbackOrder
+        {
+            get { return 0; }
+        }
+
+        ///<summary>
+        /// Initializes temporary build data.
+        /// </summary>
+        public void OnPreprocessBuild(BuildReport report)
+        {
+            if (EditorUserBuildSettings.activeBuildTarget == BuildTarget.Android && EditorUserBuildSettings.buildAppBundle)
+                MoveDataForAppBundleBuild();
+            else
+                MoveDataToDefaultLocation();
+        }
+
+        /// <summary>
+        /// Move custom asset pack data from their build location to their App Bundle data location.
+        /// </summary>
+        public static void MoveDataForAppBundleBuild()
+        {
+            if (File.Exists(CustomAssetPackUtility.CustomAssetPacksDataEditorPath))
+            {
+                if (!Directory.Exists(Application.streamingAssetsPath))
+                    Directory.CreateDirectory(Application.streamingAssetsPath);
+                File.Move(CustomAssetPackUtility.CustomAssetPacksDataEditorPath, CustomAssetPackUtility.CustomAssetPacksDataRuntimePath);
+                File.Delete(CustomAssetPackUtility.CustomAssetPacksDataEditorPath + ".meta");
```

Comment:
If it exists in AssetDatabase, it will be better to use AssetDatabase methods to move it. This will ensure it keeps the guids the same (probably doesn't have any issue, but will prevent any possible issues). This keeps the meta and the database in sync. As well as preventing the need for the database to import the "new" files at the new location.
I believe ADB.IsValidFolder will only return true is IsValidFolder(CustomAssetPackUtility.CustomAssetPacksDataEditorPath) has been imported. If false then you can use the standard File api.

Reply:
Good point! I can replace the File.Delete with AB.DeleteAsset. It looks like streaming assets aren't in the AB (AB.IsValidFolder returns false), so the File.Move will have the stay

Reply:
Ok cool, StreamingAssets are imported still. So it must be at a point in the process before an AssetDatabase.Refresh occurs.

Reply:
Hmm it seems like ADB.DeleteAsset doesn't actually delete lone .meta files. I can wrap the logic in a ADB.Start/StopAssetEditing instead. That should help with the asset importing/refreshing